deals/store.go

```diff
@@ -0,0 +1,195 @@
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []DealRecord
+	for r := range res.Next() {
+		var dr DealRecord
+		if err := json.Unmarshal(r.Value, &dr); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, dr)
+	}
+	sort.Slice(ret, func(i, j int) bool { return ret[i].Time < ret[j].Time })
+	return ret, nil
+}
+
+func (s *store) putFinalDeal(dr DealRecord) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	err := s.ds.Delete(makePendingDealKey(dr.DealInfo.ProposalCid))
+	if err != nil {
+		return fmt.Errorf("deleting PendingDeal: %s", err)
+	}
+	buf, err := json.Marshal(dr)
+	if err != nil {
+		return fmt.Errorf("marshaling DealRecord: %s", err)
+	}
+	if err := s.ds.Put(makeFinalDealKey(dr.DealInfo.ProposalCid), buf); err != nil {
+		return fmt.Errorf("put DealRecord: %s", err)
+	}
+	return nil
+}
+
+func (s *store) getFinalDeals() ([]DealRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	q := query.Query{Prefix: dsBaseStorageFinal.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []DealRecord
+	for r := range res.Next() {
+		var dealRecord DealRecord
+		if err := json.Unmarshal(r.Value, &dealRecord); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, dealRecord)
+	}
+	sort.Slice(ret, func(i, j int) bool {
+		if ret[i].DealInfo.ActivationEpoch < ret[j].DealInfo.ActivationEpoch {
+			return true
+		}
+		if ret[i].DealInfo.ActivationEpoch > ret[j].DealInfo.ActivationEpoch {
+			return false
+		}
+		return ret[i].Time < ret[j].Time
+	})
+	return ret, nil
+}
+
+func (s *store) putRetrieval(rr RetrievalRecord) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	buf, err := json.Marshal(rr)
+	if err != nil {
+		return fmt.Errorf("marshaling RetrievalRecord: %s", err)
+	}
+	if err := s.ds.Put(makeRetrievalKey(rr), buf); err != nil {
+		return fmt.Errorf("put RetrievalRecord: %s", err)
+	}
+	return nil
+}
+
+func (s *store) getRetrievals() ([]RetrievalRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	q := query.Query{Prefix: dsBaseRetrieval.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []RetrievalRecord
+	for r := range res.Next() {
+		var rr RetrievalRecord
+		if err := json.Unmarshal(r.Value, &rr); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, rr)
+	}
+	sort.Slice(ret, func(i, j int) bool {
+		return ret[i].Time < ret[j].Time
+	})
+	return ret, nil
+}
+
+func (s *store) close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	err := s.ds.Close()
+	return err
+}
```

Comment:
Important: This `close` method should be deleted. The underlying datastore will be closed by main, since what this store is seeing is just a slice of the namespace. If this was a datastore created and dedicated to this datastore, then  this would be OK.

Reply:
Ah good call.